Replace doclinter with sphinx-lint

JulienPalard opened the pull request:
Subject: Just ran [sphinx-lint](https://github.com/sphinx-contrib/sphinx-lint) on Sphinx own doc.

Unsurprisingly I found more issues in `sphinx-lint` than in `Sphinx doc` while doing so ;-)

But it found a real one:
```diff
-   For Sphinx version 3.5.1 beta 3 this would be `(3, 5, 1, 'beta', 3)``.
+   For Sphinx version 3.5.1 beta 3 this would be ``(3, 5, 1, 'beta', 3)``.
```

so I'm happy. Other things were trailing whitespaces, and so on.

<!--
  Before posting a pull request, please choose a appropriate branch:

  - Breaking changes: master
  - Critical or severe bugs: X.Y.Z
  - Others: X.Y

  For more details, see https://www.sphinx-doc.org/en/master/internals/release-process.html#branch-model
-->

JulienPalard committed:
Passing sphinx-lint on sphinx doc.

AA-Turner reviewed (approved):
Is `sphinx-lint` at a point where we can add it to CI for core Sphinx?

A

JulienPalard commented:
> Is `sphinx-lint` at a point where we can add it to CI for core Sphinx?

Yes, multiple projects already does it (cpython, pandas, sympy, ...) but I would recommend not having it mandatory yet for Sphinx, as it's typically here that we can find surprising things like rst-in-rst ;)

Do you want me to add the github workflow in the same PR?

AA-Turner commented:
> Do you want me to add the github workflow in the same PR?

Please, currently Sphinx uses a [custom linter](https://github.com/sphinx-doc/sphinx/blob/5.x/utils/doclinter.py) via tox:

https://github.com/sphinx-doc/sphinx/blob/5eeeb9ce375e2f35807b54f17cec64696a7828d7/tox.ini#L88-L95

I think using a standard tool would be beneficial.

A

JulienPalard committed:
Use sphinx-lint in Makefile, tox (and github actions) checks.

JulienPalard commented:
Updated the Makefile:

```
$ make doclinter
sphinx-lint CHANGES *.rst doc/
No problems found.
```

Updated tox.ini:

```
$ tox -e docslint
docslint create: /home/mdk/clones/sphinx/.tox/docslint
docslint develop-inst: /home/mdk/clones/sphinx
[...]
docslint run-test-pre: PYTHONHASHSEED='2544686718'
docslint run-test: commands[0] | sphinx-lint CHANGES CONTRIBUTING.rst README.rst doc/
No problems found.
_____________________________________________________________________________________________________ summary _____________________________________________________________________________________________________
  docslint: commands succeeded
  congratulations :)
```

And Github Actions side I just bumped the Python version from 3.6 to 3.7 in the lint workflow, because sphinx-lint requires it.

AA-Turner commented on .github/workflows/lint.yml:
```diff
@@ -15,7 +15,7 @@ jobs:
       uses: actions/setup-python@v1
       with:
-        python-version: 3.6
+        python-version: 3.7
```
I'd just run on the latest Python, saves updating every year for EOL dates
```suggestion
        python-version: 3
```

JulienPalard replied:
Modified to `>=3.7` just in case there exist runners with 3.6 would be picked by default.

tk0miya commented on utils/doclinter.py:
```diff
@@ -1,77 +0,0 @@
-import sys
-from typing import List
-
-MAX_LINE_LENGTH = 85
```
It seems `sphinx-lint` does not check the length of lines. So I'd not like to remove this script from our repository. How about using both?

AA-Turner replied:
Perhaps `sphinx-lint` could add a line-length option?

Otherwise we could simplify to e.g.:

```python
"""Check line length in rst documents."""

import os
import re
import sys

MAX_LINE_LENGTH = 85


def check_line_length(path: str) -> bool:
    with open(path) as f:
        document = f.readlines()

    error = False
    for i, line in enumerate(document):
        if len(line) > MAX_LINE_LENGTH:
            if not (
                re.match(r'^\s*\.\. ', line)  # ignore directives and hyperlink targets
                or re.match(r'^\s*__ ', line)  # ignore anonymous hyperlink targets
                or re.match(r'^\s*``[^`]+``$', line)  # ignore a very long literal string
            ):
                print(f'{path}:{i + 1}: the line is too long ({len(line)} > {MAX_LINE_LENGTH}).')
                error = True

    return error


if __name__ == '__main__':
    error = False
    for path in sys.argv[1:]:
        if os.path.isfile(path):
            error ^= check_line_length(path)
        elif os.path.isdir(path):
            for root, _dirs, files in os.walk(path):
                for filename in files:
                    if filename.endswith('.rst'):
                        path = os.path.join(root, filename)
                        error ^= check_line_length(path)
    raise SystemExit(error + 0)
```

A

tk0miya replied:
I'd not like to reduce our checks for docs. So it's very nice if sphinx-lint support it. Until then, it's also okay to remove duplicated features from doclinter.py.

JulienPalard replied:
Ohhh yes good catch.

We do have this checker, but it's not enabled by default:

https://github.com/sphinx-contrib/sphinx-lint/blob/b25d09c5a0cf15030c575d18078d815724a1e3e0/sphinxlint.py#L316

And if we enable it using `--severity 0` it'll whine about other things that you typically allow like using default roles.

We'll have to implement a way to enable/disable checks by name instead of by severity level for more granularity.

I'll be offline the next week, but there's no rush anyway.

JulienPalard replied:
Fixed using `--enable line-too-long --max-line-length 85`.

JulienPalard committed:
sphinx-lint: Enable line-too-long checker.

AA-Turner commented on .github/workflows/lint.yml:
```diff
@@ -13,9 +13,9 @@ jobs:
     - uses: actions/checkout@v2
     - name: Set up Python
-      uses: actions/setup-python@v1
+      uses: actions/setup-python@v3
```
```suggestion
      uses: actions/setup-python@v4
```

AA-Turner reviewed (approved):
One note, looks good otherwise.

A

tk0miya reviewed (approved):
Looks good. Thank you for developing great tool!

AA-Turner committed:
Use setup-python v4

AA-Turner commented on tox.ini:
```diff
@@ -81,9 +81,9 @@ basepython = python3
+    lint
 commands =
-    python utils/doclinter.py CHANGES CONTRIBUTING.rst README.rst doc/
+    sphinx-lint --enable line-too-long --max-line-length 85 CHANGES CONTRIBUTING.rst README.rst doc/
```
```suggestion
    sphinx-lint --disable missing-space-after-literal --enable line-too-long --max-line-length 85 CHANGES CONTRIBUTING.rst README.rst doc/
```

AA-Turner committed:
Disable ` missing-space-after-literal`
